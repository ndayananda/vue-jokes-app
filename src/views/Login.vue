<template>
    <div class="login">
        <div class="columns is-mobile is-centered">
            <div class="column is-half">
                <form class="login__form">
                    <h1 class="login__form-title is-size-4 pb-2">Login</h1>
                    <div class="login__error-msg" v-if="err.length">{{ err }}</div>
                    <div class="field">
                        <label class="label">Username/Email</label>
                        <div class="control">
                            <input class="input login__form-email-input" type="text" v-model="username"
                                placeholder="Your username or e-mail address"
                                @blur="validateUserName"
                                :class="{
                                    'is-danger': isUserNameInvalid,
                                    'is-success': !isUserNameInvalid
                                    }">
                        </div>
                        <p class="login__form-email-error help is-danger" :class="{'is-hidden': !isUserNameInvalid }">This username/email is invalid</p>
                    </div>
                    <div class="field">
                        <label class="label">Password</label>
                        <div class="control">
                            <input class="input login__form-password-input" type="password" v-model="password" placeholder="Password"
                                @blur="validatePassword"
                                :class="{
                                    'is-danger': isPasswordInvalid,
                                    'is-success': !isPasswordInvalid
                                }">
                        </div>
                        <p class="login__form-password-error help is-danger" :class="{'is-hidden': !isPasswordInvalid }" v-for="error in passwordErrors" :key="error.errorMsg">{{ error.errorMsg }}</p>
                    </div>
                    <div class="field">
                        <p class="control">
                            <button class="button is-primary" :class="{'login__form-button--disabled': isInvalid}" :disabled="isInvalid" @click="loginClickhandler">Login</button>
                        </p>
                    </div>
                </form>
            </div>
        </div>
        <BaseLoader :is-active="loginProgress" />
    </div>
</template>

<script>
import { Actions } from '@/store';
import BaseLoader from '@/components/shared/BaseLoader.vue';

const validationRules = [{
        errorMsg: 'Passwords must include one increasing straight of at least three letters, like "abc", "cde", "fgh"',
        regexp: /(?:abc|bcd|cde|def|efg|fgh|ghi|hij|ijk|jkl|klm|lmn|mno|nop|opq|pqr|qrs|rst|stu|tuv|uvw|vwx|wxy|xyz)/
    }, {
        errorMsg: 'Passwords may not contain the letters i, O, or l',
        regexp: /^[^i^o^l]+$/
    }, {
        errorMsg: 'Passwords must contain at least two non-overlapping pairs of letters, like aa, bb, or cc',
        regexp: /[a-z]*([a-z])\1([a-z])+/
    }, {
        errorMsg: 'Passwords cannot be longer than 32 characters',
        regexp: /^\w{1,32}$/
    }];

export default {
    components: {
        BaseLoader
    },
    data() {
        return {
            username: '',
            password: '',
            loginProgress: false,
            err: '',
            isUserNameInvalid: false,
            isPasswordInvalid: false,
            passwordErrors: []
        };
    },
    computed: {
        isInvalid() {
            return this.isUserNameInvalid || this.isPasswordInvalid || Boolean(!this.username) || Boolean(!this.password);
        }
    },
    methods: {
        async loginClickhandler(event) {
            event.preventDefault();
            this.loginProgress = true;
            this.err = '';
            try {
                await this.$store.dispatch(Actions.LOGIN, {username: this.username, password: this.password});
                this.loginProgress = false;
                this.$router.push('/jokes');
            }
            catch(err) {
                this.err = 'Login failed. Please check the Username & Password';
                this.loginProgress = false;
            }
        },
        validateUserName() {
            this.isUserNameInvalid =  !(/^(?=[a-z0-9.]{3,20}$)[a-z0-9]+\.?[a-z0-9]+$|^.*@\w+\.[\w.]+$/.test(this.username));
        },
        validatePassword() {
            const errors = validationRules.filter((item)=> !item.regexp.test(this.password));
            this.isPasswordInvalid = Boolean(errors.length);
            this.passwordErrors = errors.length ? errors : [];
        }
    }
}
</script>

<style lang="scss" scoped>
.login {
    min-height: calc(100vh - 230px);

    &__form {
        padding: 15px 25px;

        &-title {
            border-bottom: 1px solid lighten($color: $black, $amount: 95%);
            margin-bottom: 25px;
        }
    }
    &__error-msg {
        margin: 0 0 15px;
        font-size: 14px;
        color: $error;
    }
}
</style>