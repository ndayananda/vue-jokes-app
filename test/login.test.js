import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import flushPromises from 'flush-promises';
import '@testing-library/jest-dom'
import Login from './../src/views/Login.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

function createStore(storeConfig) {
    const defaultConfig = {

    };

    return new Vuex.Store(Object.assign(defaultConfig, storeConfig));
}

function createWrapper(options) {
    const defaultOptions = {
        localVue,
        mocks: {
            $router: {
                push: jest.fn()
            }
        },
        store: createStore()
    };

    return mount(Login, Object.assign(defaultOptions, options));
}

describe('Login View', () => {
    // Now mount the component and you have the wrapper
    let wrapper;

    beforeEach(() => wrapper = createWrapper());

    it('renders the correct login markup', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('has a login button', () => {
      expect(wrapper.find('button').exists()).toBe(true);
    });

    it('should return invalid on click of login button without entering the email or password', () => {
        const button = wrapper.find('button');
        button.trigger('click');
        expect(wrapper.vm.isInvalid).toBeTruthy();
    });

    it('will thorw invalid email error', async () => {
        await wrapper.setData({username: 'test@'});

        await wrapper.find('.login__form-email-input').trigger('blur');

        expect(wrapper.vm.isUserNameInvalid).toBeTruthy();
        expect(wrapper.element.querySelector('.login__form-email-error')).toBeVisible()
    });

    it('is a valid username/email', async () => {
        await wrapper.setData({username: 'test@test.com'});

        await wrapper.find('.login__form-email-input').trigger('blur');

        expect(wrapper.vm.isUserNameInvalid).not.toBeTruthy();
    });

    it('will thorw invalid password error', async () => {
        await wrapper.setData({password: 'welcome123'});

        await wrapper.find('.login__form-password-input').trigger('blur');

        expect(wrapper.vm.isPasswordInvalid).toBeTruthy();
        expect(wrapper.element.querySelector('.login__form-password-error')).toBeVisible()
    });

    it('is a valid password', async () => {
        await wrapper.setData({password: 'abcStraatee123'});

        await wrapper.find('.login__form-password-input').trigger('blur');

        expect(wrapper.vm.isUserNameInvalid).not.toBeTruthy();
    });

    it('After successfull login route to /jokes', async () => {
        expect.assertions(3);

        const auth = {
            namespaced: true,
            state: {},
            actions: {
                login: jest.fn().mockResolvedValue()
            }
        };

        const store = createStore({ modules: { auth }});

        const wrapper = createWrapper({store});
        await wrapper.setData({ username: 'admin', password: 'abcStraatee123'});

        await wrapper.find('button').trigger('click');

        await flushPromises();

        expect(auth.actions.login).toHaveBeenCalled();
        expect(wrapper.vm.$router.push).toHaveBeenCalled();
        expect(wrapper.vm.$router.push).toHaveBeenCalledWith('/jokes');
    });
  })