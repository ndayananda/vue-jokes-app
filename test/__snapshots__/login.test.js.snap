// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login View renders the correct login markup 1`] = `
"<div class=\\"login\\">
  <div class=\\"columns is-mobile is-centered\\">
    <div class=\\"column is-half\\">
      <form class=\\"login__form\\">
        <h1 class=\\"login__form-title is-size-4 pb-2\\">Login</h1>
        <!---->
        <div class=\\"field\\"><label class=\\"label\\">Username/Email</label>
          <div class=\\"control\\"><input type=\\"text\\" placeholder=\\"Your username or e-mail address\\" class=\\"input login__form-email-input is-success\\"></div>
          <p class=\\"login__form-email-error help is-danger is-hidden\\">This username/email is invalid</p>
        </div>
        <div class=\\"field\\"><label class=\\"label\\">Password</label>
          <div class=\\"control\\"><input type=\\"password\\" placeholder=\\"Password\\" class=\\"input login__form-password-input is-success\\"></div>
        </div>
        <div class=\\"field\\">
          <p class=\\"control\\"><button disabled=\\"disabled\\" class=\\"button is-primary login__form-button--disabled\\">Login</button></p>
        </div>
      </form>
    </div>
  </div>
  <!---->
</div>"
`;
